### PR TITLE
kcl: 0.8.9 -> 0.9.0

### PR DESCRIPTION
**Pull request by selfuryon**
## Description of changes

Update to whole kcl (kcl, kclvm and kclvm_cli) to a new release v0.9.0

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

CC @Peefy

### CODE REVIEW

**Comment by YvanDaSilva** [comment]
Hello @selfuryon 

FYI: this will not build for darwin arch.
I was also creating a patch for nixpkgs: https://github.com/NixOS/nixpkgs/compare/master...YvanDaSilva:nixpkgs:kcl-update

but now I'm stuck on :
```
'/private/tmp/nix-build-kclvm_cli-0.9.0.drv-0/source/cli/target/aarch64-apple-darwin/release/libkclvm_cli_cdylib.dylib' (no such file), '/nix/store/dq1ghaifl9rs3hz3j6v7glgkvjs404qw-rustc-1.78.0/lib/rustlib/aarch64-apple-darwin/lib/libkclvm_cli_cdylib.dylib' (no such file), '/homeless-shelter/lib/libkclvm_cli_cdylib.dylib' (no such file), '/usr/local/lib/libkclvm_cli_cdylib.dylib' (no such file), '/usr/lib/libkclvm_cli_cdylib.dylib' (no such file, not in dyld cache)
error: test failed, to rerun pass `--bin kclvm_cli`
```
You can see more on the slack kcl release thread https://cloud-native.slack.com/archives/C05TC96NWN8/p1720185862841689

File: pkgs/by-name/kc/kclvm/package.nix [hunk 1]
```diff
@@ -7,27 +7,26 @@
...
     outputHashes = {
       "inkwell-0.2.0" = "sha256-JxSlhShb3JPhsXK8nGFi2uGPp8XqZUSiqniLBrhr+sM=";
+      "protoc-bin-vendored-3.1.0" = "sha256-RRqpPMJygpKGG5NYzD93iy4htpVqFhYMmfPgbRtpUqg=";
     };
   };
 
   nativeBuildInputs = [ pkg-config protobuf ];
 
-  patches = [ ./enable_protoc_env.patch ];
-
```
Comment (YvanDaSilva) [inline pkgs/by-name/kc/kclvm/package.nix:30]: Is the patch no longer needed ?
I updated it on my branch. https://github.com/NixOS/nixpkgs/compare/master...YvanDaSilva:nixpkgs:kcl-update

Comment (YvanDaSilva) [reply pkgs/by-name/kc/kclvm/package.nix:30]: @selfuryon @Peefy 
Feel free to have a look, I updated https://github.com/NixOS/nixpkgs/compare/master...YvanDaSilva:nixpkgs:kcl-update

But I don't think it does what it is supposed to do. Check the slack thread for more info.
TLDR: It produces a kcl file of 453B, something is borked.
kclvm and kclvm_cli seem ok though.
It's the compilation of the go program that is not working as intended, I had to:
```
    export HOME=$(mktemp -d)
```

in post install. Otherwise it would panic:
```


goroutine 1 [running]:

kcl-lang.io/cli/cmd/kcl/commands.NewModCmd()

        kcl-lang.io/cli/cmd/kcl/commands/mod.go:49 +0x350

kcl-lang.io/cli/cmd/kcl/commands.New()

        kcl-lang.io/cli/cmd/kcl/commands/root.go:119 +0x238

main.main()

        kcl-lang.io/cli/cmd/kcl/main.go:14 +0x1c

panic: failed to load config file '' for kpm.

internal bugs, failed to load working directory

could not access $KCL_PKG_PATH.

internal bugs, failed to create directory

mkdir /homeless-shelter: read-only file system

```

Comment (YvanDaSilva) [reply pkgs/by-name/kc/kclvm/package.nix:30]: Made the PR HERE: https://github.com/NixOS/nixpkgs/pull/327390

It works on both nixos x86 and darwin aarch

**Comment by selfuryon** [comment]
It's better PR: https://github.com/NixOS/nixpkgs/pull/327390
